File: python/stacked.py
#!/usr/bin/python3

import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
#import os

minYear=1900

print('Loading data')
movies = pd.read_json('../data/sorted.json').sort_values('year')
    
print('Loaded data')
#find min
for year in movies['year']:
  if year > minYear:
    min = year
    break
    
years = list(np.arange(min, movies['year'].max()+1))
#print(years)
print('Getting unique genres')
genres = []
for row in movies['genres']:
  for genre in row:
    if genre not in genres:
      genres.append(genre)
          
#print(genres)
print('Counting genres')
data = np.zeros((len(years), len(genres)))

count=0
for index, row in movies.iterrows():
  if row['year'] > minYear:
    for genre in row['genres']:
      data[years.index(row['year']), genres.index(genre)] += 1
    count+=1
    print(f'{count}/{len(movies)}', end='\r')      
      
np.save('../data/stacked.npy', data)
dpi = 128
plt.figure(figsize=(2560/dpi, 1440/dpi), dpi=dpi)
print('Generating stacked plot')
plt.stackplot(years, data.T)
plt.legend(genres, loc=2)
#plt.show()
plt.savefig('stackedPlot.png')

print('Normalizing data')
for i in range(len(data)):
  data[i] /= datap[i].sum()
  print(f'{i}/{len(data)}')

File: R/stacked.r
#!/usr/bin/R -s -f

library("rjson")
library(jsonlite)
library(ggplot2)
library(dplyr)

# json to dataframe
myData <- fromJSON("myJSON.json")

minYear = 1990
maxYear = max(myData$year)
print(maxYear)
years = seq(minYear, maxYear, 1)
genres <- unlist(myData$genres)
genres <- unique(genres)
print(genres)
# initializing a matrix of 0s 
mat = matrix(0, length(years), length(genres))
count = 0
ig = 1
iy = 1

for(i in 1:length(myData$year)){
  iy <- which(years==myData$year[i])
  print(myData$year[i])
  if(myData$year[i] >= minYear){
    for(genre in myData$genres[i]){
      for(g in genre){
        ig <- which(genres==g)
        mat[iy, ig]  <-  mat[iy, ig] + 1
      }  
    }
  }
}
a=c(mat)
# create data
time <- as.numeric(years)  # x Axis
value <- a             # y Axis
group <- genres       # group, one shape per group
data <- data.frame(time, value, group)
# create and save plot
ggplot(data, aes(x=time, y=value, fill=group)) + geom_area(alpha=0.6 , size=0.1, colour="black")
ggsave("stacked.png", width = 2560, height = 1440, units = "px")
# normalize plot
for (i in 1:nrow(mat)){
  print(mat[i, ])
  mat[i, ] <- mat[i, ] / sum(mat[i, ])
}
a=c(mat)
# create data
time <- as.numeric(years)  # x Axis
value <- a             # y Axis
group <- genres       # group, one shape per group
data <- data.frame(time, value, group)
# create and save plot
ggplot(data, aes(x=time, y=value, fill=group)) + geom_area(alpha=0.6 , size=0.1, colour="black")
ggsave("normalizedStacked.png", width = 2560, height = 1440, units = "px")